sql/core/src/test/scala/org/apache/spark/sql/streaming/test/DataStreamReaderWriterSuite.scala

```diff
@@ -532,4 +532,50 @@ class DataStreamReaderWriterSuite extends StreamTest with BeforeAndAfter {
     assert(e.getMessage.contains("does not support recovering"))
     assert(e.getMessage.contains("checkpoint location"))
   }
+
+  test("SPARK-18510: Data corruption from user specified partition column schemas") {
+    import org.apache.spark.sql.functions.udf
+    import testImplicits._
+    withTempDir { src =>
+      val createArray = udf { (length: Long) =>
+        for (i <- 1 to length.toInt) yield i.toString
+      }
+      spark.range(4).select(createArray('id + 1) as 'ex, 'id, 'id % 4 as 'part).coalesce(1).write
+        .partitionBy("part", "id")
+        .mode("overwrite")
+        .parquet(src.toString)
+      // Specify a random ordering of the schema, partition column in the middle, etc.
+      // Also let's say that the partition columns are Strings instead of Longs.
+      // partition columns should go to the end
+      val schema = new StructType()
+        .add("id", StringType)
+        .add("ex", ArrayType(StringType))
+
+      try {
+        val sdf = spark.readStream
+          .schema(schema)
+          .format("parquet")
+          .load(src.toString)
+
+        assert(sdf.schema.toList === List(
+          StructField("ex", ArrayType(StringType)),
+          StructField("part", IntegerType), // inferred partitionColumn dataType
+          StructField("id", StringType))) // used user provided partitionColumn dataType
+
+        val sq = sdf.writeStream
+          .queryName("corruption_test")
+          .format("memory")
+          .start()
+        sq.processAllAvailable()
+        checkAnswer(
+          spark.table("corruption_test"),
+          // notice how `part` is ordered before `id`
+          Row(Array("1"), 0, "0") :: Row(Array("1", "2"), 1, "1") ::
+            Row(Array("1", "2", "3"), 2, "2") :: Row(Array("1", "2", "3", "4"), 3, "3") :: Nil
+        )
+      } finally {
+        spark.streams.active.foreach(_.stop())
```

Comment:
nit: you can just to sq.stop() instead of try...catch, because `after { }`of this testsuite already has `spark.streams.active.foreach(_.stop())`